File: frontend/src/api/assessment.ts
import api from './index';
import type { AssessmentItem, AssessmentResults } from '../types';

export interface AssessmentItemsResponse {
  items: AssessmentItem[];
  totalItems: number;
  currentIndex: number;
  responses: Record<string, string>;
  title: string;
}

export interface SubmitResponse {
  success: boolean;
  nextIndex: number;
  isComplete: boolean;
}

export const getAssessmentItems = async (): Promise<AssessmentItemsResponse> => {
  const response = await api.get<AssessmentItemsResponse>('/assessment/items');
  return response.data;
};

export const submitAssessmentResponse = async (
  itemId: string,
  response: string
): Promise<SubmitResponse> => {
  const res = await api.post<SubmitResponse>('/assessment/submit', {
    itemId,
    response,
  });
  return res.data;
};

export const skipAssessmentQuestion = async (itemId: string): Promise<SubmitResponse> => {
  const response = await api.post<SubmitResponse>('/assessment/skip', { itemId });
  return response.data;
};

export const getAssessmentResults = async (): Promise<AssessmentResults> => {
  const response = await api.get<{
    success: boolean;
    framework?: string;
    results: {
      framework?: string;
      band_scale?: number;
      global_stage: number;
      domain_bands: Record<string, number>;
      proficiency_level?: string;
      proficiency_description_en?: string;
      actfl_level?: string;
      actfl_description_en?: string;
    };
    proficiencyLevel?: string;
    proficiencyDescription?: string;
    actflLevel?: string;
    actflDescription?: string;
    sklcLevel?: string;
    sklcDescription?: string;
  }>('/assessment/results');

  const proficiencyLevel =
    response.data.proficiencyLevel ||
    response.data.actflLevel ||
    response.data.results.proficiency_level ||
    response.data.results.actfl_level ||
    response.data.sklcLevel ||
    '';

  const proficiencyDescription =
    response.data.proficiencyDescription ||
    response.data.actflDescription ||
    response.data.results.proficiency_description_en ||
    response.data.results.actfl_description_en ||
    response.data.sklcDescription ||
    '';

  return {
    framework: response.data.framework || response.data.results.framework || 'ACTFL',
    bandScale: response.data.results.band_scale,
    globalStage: response.data.results.global_stage,
    domainBands: response.data.results.domain_bands,
    proficiencyLevel,
    proficiencyDescription,
    actflLevel: response.data.actflLevel || response.data.results.actfl_level || proficiencyLevel,
    actflDescription:
      response.data.actflDescription || response.data.results.actfl_description_en || proficiencyDescription,
    sklcLevel: response.data.sklcLevel || proficiencyLevel,
    sklcDescription: response.data.sklcDescription || proficiencyDescription,
  };
};

export const resetAssessment = async (): Promise<void> => {
  await api.post('/assessment/reset');
};

export const updateCategories = async (categories: string[]): Promise<void> => {
  await api.post('/categories', { categories });
};

File: backend/routes/assessment.py
from flask import Blueprint, jsonify, request

from backend.route_deps import RouteDeps


def create_assessment_blueprint(deps: RouteDeps) -> Blueprint:
    bp = Blueprint('assessment_routes', __name__)

    @bp.route('/api/assessment/status')
    @deps.login_required
    def api_assessment_status():
        uid = deps.get_current_user_uid()
        assessment_data = deps.get_assessment()
        assessment_state = deps.db.get_assessment_state(uid) or {}
        return jsonify({
            'current_index': assessment_state.get('current_item_index', 0),
            'total_items': len(assessment_data['items']),
            'responses_count': len(assessment_state.get('responses', {})),
        })

    @bp.route('/api/assessment/items', methods=['GET'])
    @deps.login_required
    def api_assessment_items():
        """Get all assessment items and current progress (JSON API)."""
        uid = deps.get_current_user_uid()
        assessment_data = deps.get_assessment()

        assessment_state = deps.db.get_assessment_state(uid) or {}
        current_index = assessment_state.get('current_item_index', 0)
        responses = assessment_state.get('responses', {})

        return jsonify({
            'items': assessment_data['items'],
            'totalItems': len(assessment_data['items']),
            'currentIndex': current_index,
            'responses': responses,
            'title': assessment_data['title'],
        })

    @bp.route('/api/assessment/submit', methods=['POST'])
    @deps.login_required
    def api_assessment_submit_json():
        """Submit an assessment response (JSON API)."""
        uid = deps.get_current_user_uid()
        data = request.get_json() or {}

        item_id = data.get('itemId')
        response = data.get('response', '')

        if not item_id:
            return jsonify({'success': False, 'error': 'Item ID is required'}), 400

        assessment_data = deps.get_assessment()
        items = assessment_data['items']

        current_index = next((i for i, item in enumerate(items) if item['id'] == item_id), None)
        if current_index is None:
            return jsonify({'success': False, 'error': 'Invalid item ID'}), 400

        deps.db.update_assessment_response(uid, item_id, response, current_index + 1)
        is_complete = (current_index + 1) >= len(items)

        return jsonify({'success': True, 'nextIndex': current_index + 1, 'isComplete': is_complete})

    @bp.route('/api/assessment/skip', methods=['POST'])
    @deps.login_required
    def api_assessment_skip_json():
        """Skip current assessment question (JSON API)."""
        uid = deps.get_current_user_uid()
        data = request.get_json() or {}

        item_id = data.get('itemId')
        if not item_id:
            return jsonify({'success': False, 'error': 'Item ID is required'}), 400

        assessment_data = deps.get_assessment()
        items = assessment_data['items']

        current_index = next((i for i, item in enumerate(items) if item['id'] == item_id), None)
        if current_index is None:
            return jsonify({'success': False, 'error': 'Invalid item ID'}), 400

        deps.db.update_assessment_response(uid, item_id, '', current_index + 1)
        is_complete = (current_index + 1) >= len(items)

        return jsonify({'success': True, 'nextIndex': current_index + 1, 'isComplete': is_complete})

    @bp.route('/api/assessment/results', methods=['GET'])
    @deps.login_required
    def api_assessment_results_json():
        """Get computed assessment results (JSON API)."""
        uid = deps.get_current_user_uid()

        results = deps.db.get_assessment_results(uid)
        if not results:
            assessment_state = deps.db.get_assessment_state(uid)
            if assessment_state:
                responses = assessment_state.get('responses', {})
                if responses:
                    assessment_data = deps.get_assessment()
                    results = deps.compute_results(assessment_data, responses)
                    deps.db.save_assessment_results(uid, results)

        if results:
            global_stage = results.get('global_stage', 0)
            sklc_info = deps.get_sklc_description(global_stage)

            return jsonify({
                'success': True,
                'results': results,
                'sklcLevel': sklc_info['level'],
                'sklcDescription': sklc_info['description'],
            })

        return jsonify({'success': False, 'error': 'No results available'}), 404

    @bp.route('/api/assessment/reset', methods=['POST'])
    @deps.login_required
    def api_assessment_reset_json():
        """Reset assessment progress (JSON API)."""
        uid = deps.get_current_user_uid()
        deps.db.reset_assessment(uid)
        return jsonify({'success': True})

    @bp.route('/api/categories', methods=['POST'])
    @deps.login_required
    def api_update_categories():
        """Update selected practice categories (JSON API)."""
        uid = deps.get_current_user_uid()
        data = request.get_json() or {}

        categories = data.get('categories', [])

        db_results = deps.db.get_assessment_results(uid)
        if not db_results:
            assessment_state = deps.db.get_assessment_state(uid)
            if assessment_state:
                responses = assessment_state.get('responses', {})
                if responses:
                    assessment_data = deps.get_assessment()
                    results = deps.compute_results(assessment_data, responses)
                    deps.db.save_assessment_results(uid, results)

        deps.db.update_selected_categories(uid, categories)
        return jsonify({'success': True, 'categories': categories})

    return bp
